### PR TITLE
Re-add fix for `accurateCastOrNull()`

### LINKED ISSUE #38585: accurateCastOrNull raises an exception, rather than simply returning null when casting an invalid value to Bool
**Describe what's wrong**

accurateCastOrNull raises an exception, rather than simply returning null when casting an invalid value to Bool

```
Code: 467. DB::Exception: Received from 127.0.0.1:9000. DB::Exception: Cannot parse boolean value here: 'test', should be 'true' or 'false' controlled by setting bool_true_representation and bool_false_representation or one of True/False/T/F/Y/N/Yes/No/On/Off/Enable/Disable/Enabled/Disabled/1/0: While processing accurateCastOrNull('test', 'Bool'). (CANNOT_PARSE_BOOL)
```

**Does it reproduce on recent release?**

ClickHouse server version 22.7.1.906 (official build).

**How to reproduce**

```
select accurateCastOrNull('test', 'Bool')
```

**Expected behavior**

```
I expect it to behave the same as an invalid cast to something like UInt8

SELECT accurateCastOrNull('test', 'UInt8')

Query id: e7d0473b-6ae0-455d-9eae-fb3827e7c671

┌─accurateCastOrNull('test', 'UInt8')─┐
│                                ᴺᵁᴸᴸ │
└─────────────────────────────────────┘
```

### PR DESCRIPTION
**Pull request by aiven-sal**
Fixes #38585

This is a re-submit of #54136 which was reverted with #54472.

`accurateCastOrNull` must not throw when casting an arbitrary string to Bool. This PR also fixes an unrelated issue that caused `accurateCastOrNull` to behave like `accurateCastOrDefault` on error when casting to IPv4 or IPv6.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
SQL function `accurateCastOrNull()` now 1. no longer throws an exception but instead returns `NULL` when casting an arbitrary string without "boolean equivalent" to bool, e.g. `SELECT accurateCastOrDefault('test', 'Bool')`, 2. when casting an invalid string to an IP now returns `NULL` instead of an address with all zeros, e.g. `SELECT accurateCastOrDefault('192.0.2.1x', 'IPv4')`, and, 3. with setting `cast_keep_nullable = 1`, casting a nullable empty string to an IP now produces `NULL` which is consistent with the behavior for other types.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit e7e11c75e821ded4cb8e9709a30a9afd2bee5dfa with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54629/e7e11c75e821ded4cb8e9709a30a9afd2bee5dfa/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

**Review by rschu1ze — APPROVED** [review]
Thanks. Let's wait till CI results come back and merge then.

**Comment by rschu1ze** [comment]
@aiven-sal Failure of `02811_csv_input_field_type_mismatch` is related, would you like to check?

File: src/Functions/FunctionsConversion.h [hunk 1]
```diff
@@ -1683,7 +1684,25 @@ struct ConvertImplGenericFromString
 
                 const auto & val = col_from_string->getDataAt(i);
                 ReadBufferFromMemory read_buffer(val.data, val.size);
-                serialization_from.deserializeWholeText(column_to, read_buffer, format_settings);
+                try
+                {
+                    serialization_from.deserializeWholeText(column_to, read_buffer, format_settings);
+                }
+                catch (const Exception & e)
```
Comment (rschu1ze) [inline src/Functions/FunctionsConversion.h:1691]: `ConvertImplGenericFromString` (l. 1644) is a generic handler, putting special handling only for bool types (l. 1691ff.) into it is not beautiful. I wonder how parsing errors of other data types are handled, e.g. ` select accurateCastOrNull('invalid', 'Date');`? (didn't check deeper where the logic is). A better alternative would be to handle errors in bool parsing equivalently to these types.

Comment (rschu1ze) [reply src/Functions/FunctionsConversion.h:1691]: In my previous `Date` example, which code updates the null bitmap on the result column then?

Comment (rschu1ze) [reply src/Functions/FunctionsConversion.h:1691]: Okay then.

Let's address the other comments (about tests) and then we can merge.

**Comment by rschu1ze** [comment]
AST fuzzer unfortunately complains for queries:
- `SELECT CAST(if(number % 2, 'truetruetruetrue', NULL), 'Nullable(Bool)'`
- `SELECT (3.4028234663852886e38 + 9223372036854775806) + -inf, CAST(if(number % 2, 'truetruetruetrue', NULL), 'Nullable(Bool)') FROM numbers(255)`
- `SELECT CAST(if(number % 10, 'truetruetruetruetruetruetruetruetruetruetruetruetruetruetruetrue', NULL), 'Nullable(Bool)') FROM numbers(1024)`

**Pull request merged**